kubespawner/spawner.py

```diff
@@ -726,119 +747,135 @@ def _hub_connect_port_default(self):
+            c.KubeSpawner.lifecycle_hooks = {
+                "postStart": {
+                    "exec": {
+                        "command": ["/bin/sh", "-c", "echo Hello from the postStart handler > /usr/share/message"]
+                    }
+                },
+                "preStop": {
+                    "exec": {
+                        "command": ["/usr/sbin/nginx", "-s", "quit"]
+                    }
+                }
+            }
+
+        See `the Kubernetes documentation <https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/>`_
+        for more info on what lifecycle hooks are and why you might want to use them!
         """
     )
 
     init_containers = List(
-        None,
+        [],
         config=True,
         help="""
         List of initialization containers belonging to the pod.
 
         This list will be directly added under `initContainers` in the kubernetes pod spec,
-        so you should use the same structure. Each item in the list is container configuration
-        which follows spec at https://v1-6.docs.kubernetes.io/docs/api-reference/v1.6/#container-v1-core.
-
-        One usage is disabling access to metadata service from single-user notebook server with configuration below:
-        initContainers:
-
-        .. code::yaml
-
-            - name: init-iptables
-              image: <image with iptables installed>
-              command: ["iptables", "-A", "OUTPUT", "-p", "tcp", "--dport", "80", "-d", "169.254.169.254", "-j", "DROP"]
-              securityContext:
-                capabilities:
-                  add:
-                  - NET_ADMIN
+        so you should use the same structure. Each item in the dict must a field
+        of the `V1Container specification <https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.11/#container-v1-core>`_.
+
+        One usage is disabling access to metadata service from single-user
+        notebook server with configuration below::
+
+            c.KubeSpawner.init_containers = [{
+                "name": "init-iptables",
+                "image": "<image with iptables installed>",
+                "command": ["iptables", "-A", "OUTPUT", "-p", "tcp", "--dport", "80", "-d", "169.254.169.254", "-j", "DROP"],
+                "securityContext": {
+                    "capabilities": {
+                        "add": ["NET_ADMIN"]
+                    }
+                }
+            }]
 
 
-        See https://kubernetes.io/docs/concepts/workloads/pods/init-containers/ for more
-        info on what init containers are and why you might want to use them!
+        See `the Kubernetes documentation <https://kubernetes.io/docs/concepts/workloads/pods/init-containers/>`_
+        for more info on what init containers are and why you might want to use them!
 
         To user this feature, Kubernetes version must greater than 1.6.
         """
     )
 
     extra_container_config = Dict(
-        None,
+        {},
```

Comment:
The default for container (List, Dict) traits is an empty container, so it would be more idiomatic to leave these out, rather than define them as an unused empty dict.

Reply:
Thanks for helping me grasp whats going on, I though they would have a default value of Undefined from the source code, but I realize now the logic. It makes sense to write the default value of `default_value` to be `Undefined` instead of `{}` - to avoid using a reference type value as a parameters default value just like you made me get to learn previously!

I'll exclude the positional and named way to set the default_value parameter altogether and allow the hidden-in-source initialization logic of Dict() and List() initialize them as `{}` and `[]` respectively. 

![image](https://user-images.githubusercontent.com/3837114/44029598-25bbebc4-9efe-11e8-8d0b-a778a8328659.png)

Reply:
I managed to see that Dict() initialized to `{}` after closer inspection of the source, but for List(), I fail to see how it ends up with `[]`.

![image](https://user-images.githubusercontent.com/3837114/44030509-4ac357a6-9f01-11e8-9ab4-b17b0952bbdc.png)

I don't see any trace of this in the documentation either:
https://traitlets.readthedocs.io/en/stable/trait_types.html#containers

---

@minrk After inspecting the List traitlet's source code and documentation, and still remaining confused as to how this will become an empty list I'm getting uneasy. I think we should deviate from the idiomatic way to not redundantly specify things, as it is the best way we can to better document the behavior for users and developers of KubeSpawner. Writing a comment at one place won't suffice, and writing it in all the KubeSpawner's traitlets docstrings would really clutter the documentation.

What do you think? I'll happily strip away these default_value positional arguments as long as I know you have heard this concern and still would appreciate to strip them.

Reply:
I went ahead and removed it. At least it is now systematically implicit where it previously was a mix of implicit and explicit.